src/OptimalTransport.jl

```diff
@@ -117,62 +117,114 @@ function emd2(μ, ν, C, optimizer; plan=nothing)
+    check_convergence=check_marginal_step,
+    maxiter::Int=1_000,
+)
+    if tol !== nothing
+        Base.depwarn(
+            "keyword argument `tol` is deprecated, please use `atol` and `rtol`",
+            :sinkhorn_gibbs,
+        )
     end
+    if check_marginal_step !== nothing
+        Base.depwarn(
+            "keyword argument `check_marginal_step` is deprecated, please use `check_convergence`",
+            :sinkhorn_gibbs,
+        )
+    end
+    sum(μ) ≈ sum(ν) ||
+        throw(ArgumentError("source and target marginals must have the same mass"))
+
+    # set default values of tolerances
+    T = float(Base.promote_eltype(μ, ν, K))
+    _atol = atol === nothing ? 0 : atol
+    _rtol = rtol === nothing ? (_atol > zero(_atol) ? zero(T) : sqrt(eps(T))) : rtol
 
     # initial iteration
-    temp_v = vec(sum(K; dims=2))
-    u = mu ./ temp_v
-    temp_u = K' * u
-    v = nu ./ temp_u
+    u = μ ./ sum(K; dims=2)
+    v = ν ./ (K' * u)
+    tmp1 = K * v
+    tmp2 = similar(u)
 
+    norm_μ = sum(abs, μ) # for convergence check
     isconverged = false
+    check_step = check_convergence === nothing ? 10 : check_convergence
     for iter in 0:maxiter
-        # check mu marginal
-        if iter % check_marginal_step == 0
-            mul!(temp_v, K, v)
-            @. temp_v = abs(mu - u * temp_v)
-
-            err = maximum(temp_v)
-            @debug "Sinkhorn algorithm: iteration $iter" err
+        if iter % check_step == 0
+            # check source marginal
+            # do not overwrite `tmp1` but reuse it for computing `u` if not converged
+            @. tmp2 = u * tmp1
+            norm_uKv = sum(abs, tmp2)
+            @. tmp2 = μ - tmp2
+            norm_diff = sum(abs, tmp2)
+
+            @debug "Sinkhorn algorithm (" *
+                   string(iter) *
+                   "/" *
+                   string(maxiter) *
+                   ": absolute error of source marginal = " *
+                   string(norm_diff)
 
             # check stopping criterion
-            if err < tol
+            if norm_diff < max(_atol, _rtol * max(norm_μ, norm_uKv))
+                @debug "Sinkhorn algorithm ($iter/$maxiter): converged"
                 isconverged = true
                 break
             end
         end
 
         # perform next iteration
         if iter < maxiter
-            mul!(temp_v, K, v)
-            @. u = mu / temp_v
-            mul!(temp_u, K', u)
-            @. v = nu / temp_u
+            @. u = μ / tmp1
+            mul!(v, K', u)
+            @. v = ν / v
+            mul!(tmp1, K, v)
         end
     end
 
     if !isconverged
-        @warn "Sinkhorn algorithm did not converge"
+        @warn "Sinkhorn algorithm ($maxiter/$maxiter): not converged"
     end
 
     return u, v
 end
 
 """
-    sinkhorn(μ, ν, C, ε; tol=1e-9, check_marginal_step=10, maxiter=1_000)
+    sinkhorn(
+        μ, ν, C, ε; atol=0, rtol=atol > 0 ? 0 : √eps, check_convergence=10, maxiter=1_000
+    )
 
```

Comment:
I missed this in the first read, but why `atol = 0` as the default? Why not `1e-9` as before? Also, is the `eps` in `sqrt(eps)`  the `ε` in the argument? If so, then I thinks it's clearer to have `\\sqrt(ε)`.

Reply:
Absolute tolerances depend on the scale and hence usually are meaningless. The values are the default values of `isapprox` which also contains an example that motivates this and explains why one should use `isapprox(x, y)` instead of `isapprox(x - y, 0)`. The `eps` is the machine epsilon of the floating point type (that depends on mu, nu, and C, as explained in the docstring). Also the notation with `√eps` is copied from `isapprox`. I thought it would be OK since it is part of base Julia.

Reply:
> Absolute tolerances depend on the scale and hence usually are meaningless.

IMO even more so with the inf-norm since for vectors of increasing dimension `n` even vectors with increasingly small inf-norm of `c/n`, where `c` is a constant, can have a constant 1-norm of `c`.